include/fiction/algorithms/physical_design/on_the_fly_circuit_design_on_defective_surface.hpp

```diff
@@ -0,0 +1,210 @@
+     */
+    exact_physical_design_params exact_design_parameter = {};
+};
+
+/**
+ * Statistics for the on-the-fly defect-aware circuit design.
+ */
+template <typename GateLyt>
+struct on_the_fly_circuit_design_stats
+{
+    /**
+     * The total runtime of the operational domain computation.
+     */
+    mockturtle::stopwatch<>::duration time_total{0};
+    /**
+     * The `stats` of the *exact* algorithm.
+     */
+    exact_physical_design_stats exact_stats{};
+    /**
+     * The gate-level layout after P&R.
+     */
+    std::optional<GateLyt> gate_layout{};
+};
+
+namespace detail
+{
+
+template <typename Ntk, typename CellLyt, typename GateLyt>
+class on_the_fly_circuit_design_impl
+{
+  public:
+    on_the_fly_circuit_design_impl(const Ntk& ntk, const on_the_fly_circuit_design_params<CellLyt>& design_params,
+                                   const GateLyt& tiling, on_the_fly_circuit_design_stats<GateLyt>& st) :
+            lattice_tiling{tiling},
+            network{ntk},
+            params{design_params},
+            stats{st}
+    {}
+
+    [[nodiscard]] sidb_defect_surface<CellLyt> design_circuit_on_defective_surface()
+    {
+        const mockturtle::stopwatch stop{stats.time_total};
+
+        std::optional<GateLyt> gate_level_layout = std::nullopt;
+
+        CellLyt lyt{};
+
+        // generating the blacklist based on neutral defects. The long-range electrostatic influence of charged defects
+        // is not considered as gates are designed on-the-fly.
+        auto black_list = sidb_surface_analysis<sidb_skeleton_bestagon_library>(
+            lattice_tiling, params.parameterized_gate_library_parameters.defect_surface, std::make_pair(0, 0));
+
+        while (!gate_level_layout.has_value())
+        {
+            // P&R with *exact* and the pre-determined blacklist
+            gate_level_layout =
+                exact_with_blacklist<GateLyt>(network, black_list, params.exact_design_parameter, &stats.exact_stats);
+
+            if (gate_level_layout.has_value())
+            {
+                try
+                {
+                    lyt = apply_parameterized_gate_library<CellLyt, parameterized_gate_library, GateLyt,
+                                                           parameterized_gate_library_params<CellLyt>>(
+                        *gate_level_layout, params.parameterized_gate_library_parameters);
+                }
+
+                // on-the-fly gate design was unsuccessful at a certain tile. Hence, this tile-gate pair is added to the
+                // blacklist and the process is rerun.
+                catch (const gate_design_exception<tt, GateLyt>& e)
+                {
+                    black_list[e.which_tile()][e.which_truth_table()].push_back(e.which_port_list());
+                }
+
+                catch (const std::exception& e)
+                {
+                    std::cerr << e.what() << '\n';
+                }
```

Comment:
Then what? You print the error but how is the program going to recover?

Reply:
What are the possible exceptions that can be thrown here even?

Reply:
@marcelwa sorry, this catch is superfluous.